// react-aaui/lib/modal/Modal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _keyCodes = require('../shared/keyCodes');

var _keyCodes2 = _interopRequireDefault(_keyCodes);

var _scopeTab = require('../shared/scopeTab');

var _scopeTab2 = _interopRequireDefault(_scopeTab);

var _utils = require('../shared/utils');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Modal = function (_React$PureComponent) {
  (0, _inherits3.default)(Modal, _React$PureComponent);

  function Modal() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Modal);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Modal.__proto__ || (0, _getPrototypeOf2.default)(Modal)).call.apply(_ref, [this].concat(args))), _this), _this.autoFocus = function () {
      _this.removeTimer = setTimeout(function () {
        _this.focusContent();
        _this.focusAfterRender = false;
      }, 100);
    }, _this.fixBody = function () {
      var bodyStyles = document.body && document.body.style || {};
      var scrollbarWidth = 0;
      if (document.documentElement) {
        scrollbarWidth = window.innerWidth - document.documentElement.clientWidth;
      }

      _this.bodyOverflowX = bodyStyles.overflowX;
      _this.bodyOverflowY = bodyStyles.overflowY;
      _this.bodyPaddingRight = bodyStyles.paddingRight;

      bodyStyles.overflowX = 'hidden';
      bodyStyles.overflowY = 'hidden';

      /* istanbul ignore else */
      if (scrollbarWidth > 0 && document.body) {
        var _getComputedStyle = getComputedStyle(document.body),
            paddingRight = _getComputedStyle.paddingRight;

        bodyStyles.paddingRight = parseInt(paddingRight, 10) + parseInt(scrollbarWidth, 0) + 'px';
        if (_this.root) {
          _this.root.style.paddingRight = bodyStyles.paddingRight;
        }
      }
    }, _this.unfixBody = function () {
      var bodyStyles = document.body && document.body.style || {};
      bodyStyles.overflowX = _this.bodyOverflowX;
      bodyStyles.overflowY = _this.bodyOverflowY;
      bodyStyles.paddingRight = _this.bodyPaddingRight;
      if (_this.root) {
        _this.root.style.paddingRight = '';
      }
    }, _this.focusContent = function () {
      var autoFocusClose = _this.props.autoFocusClose;

      if (!_this.contentHasFocus() && autoFocusClose) {
        _this.closeIcon && _this.closeIcon.focus();
      }
    }, _this.contentHasFocus = function () {
      return document.activeElement === _this.root || _this.root && _this.root.contains(document.activeElement);
    }, _this.handleModalKeydown = function (e) {
      switch (e.keyCode) {
        case _keyCodes2.default.ESCAPE:
          _this.props.onClose(e);
          break;
        case _keyCodes2.default.TAB:
          _this.handleTab(e);
          break;
        default:
          break;
      }
    }, _this.handleCloseIconKeydown = function (e) {
      switch (e.keyCode) {
        case _keyCodes2.default.SPACE:
          _this.props.onClose(e);
          e.preventDefault();
          break;
        default:
          break;
      }
    }, _this.bindRootRef = function (ref) {
      _this.root = ref;
    }, _this.bindBoxRef = function (ref) {
      _this.box = ref;
    }, _this.bindCloseIconRef = function (ref) {
      _this.closeIcon = ref;
    }, _this.render = function () {
      var _this$props = _this.props,
          className = _this$props.className,
          style = _this$props.style,
          title = _this$props.title,
          shown = _this$props.shown,
          onClose = _this$props.onClose,
          children = _this$props.children,
          rest = (0, _objectWithoutProperties3.default)(_this$props, ['className', 'style', 'title', 'shown', 'onClose', 'children']);

      var classes = (0, _classnames2.default)({
        modal: true,
        'is-open': shown
      }, className);

      return React.createElement(
        'div',
        (0, _extends3.default)({}, rest, { className: 'modal-wrap', role: 'dialog' }),
        React.createElement(
          'div',
          { className: classes, style: style, ref: _this.bindRootRef, onKeyDown: _this.handleModalKeydown },
          React.createElement(
            'section',
            { className: 'modal-box', ref: _this.bindBoxRef },
            React.createElement(
              'header',
              { className: 'modal-header' },
              React.createElement(
                'h3',
                { className: 'modal-title' },
                title
              ),
              React.createElement('span', { role: 'button', className: 'icon-close modal-close', tabIndex: '0', 'aria-label': 'close-dialog', ref: _this.bindCloseIconRef, onClick: onClose, onKeyDown: _this.handleCloseIconKeydown })
            ),
            children
          )
        ),
        React.createElement('div', { className: 'modal-mask' })
      );
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(Modal, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (this.props.shown) {
        this.fixBody();
        this.autoFocus();
      }
      this.focusAfterRender = true;
      this.handleTab = (0, _scopeTab2.default)(this.root);
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate(_ref2) {
      var shown = _ref2.shown;

      if (shown !== this.props.shown) {
        if (this.props.shown) {
          this.fixBody();
          this.originFocusEl = document.activeElement;
          /* istanbul ignore else */
          if (this.focusAfterRender) {
            this.autoFocus();
          }
        } else {
          this.unfixBody();
          this.originFocusEl && this.originFocusEl.focus();
          this.focusAfterRender = true;
          clearTimeout(this.removeTimer);
        }
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.unfixBody();
    }
  }]);
  return Modal;
}(React.PureComponent);

Modal.defaultProps = {
  shown: false,
  onClose: _utils.noop,
  autoFocusClose: true
};
exports.default = Modal;


Modal.displayName = 'AAUIModal';
