// react-aaui/lib/radio/RadioGroup.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _propTypes = require('prop-types');

var _Radio = require('./Radio');

var _Radio2 = _interopRequireDefault(_Radio);

var _utils = require('../shared/utils');

require('./RadioGroup.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RadioGroup = function (_React$PureComponent) {
  (0, _inherits3.default)(RadioGroup, _React$PureComponent);

  function RadioGroup(props) {
    (0, _classCallCheck3.default)(this, RadioGroup);

    var _this = (0, _possibleConstructorReturn3.default)(this, (RadioGroup.__proto__ || (0, _getPrototypeOf2.default)(RadioGroup)).call(this, props));

    _this.onChange = function (e) {
      var value = e.target.value;
      var onChange = _this.props.onChange;


      if (!('value' in _this.props)) {
        _this.setState({
          value: value
        });
      }

      onChange(e);
    };

    _this.state = {
      value: props.value || props.defaultValue
    };
    return _this;
  }

  (0, _createClass3.default)(RadioGroup, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        auiRadioGroup: {
          onChange: this.onChange,
          value: this.state.value,
          disabled: this.props.disabled,
          size: this.props.size,
          name: this.props.name
        }
      };
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if ('value' in nextProps) {
        this.setState({
          value: nextProps.value
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var data = this.props.data;
      var children = this.props.children;


      if (data) {
        children = data.map(function (_ref, index) {
          var text = _ref.text,
              value = _ref.value,
              rest = (0, _objectWithoutProperties3.default)(_ref, ['text', 'value']);
          return React.createElement(
            _Radio2.default,
            (0, _extends3.default)({ key: index, value: value }, rest),
            text
          );
        });
      }

      return React.createElement(
        'fieldset',
        null,
        React.createElement('legend', null),
        children
      );
    }
  }]);
  return RadioGroup;
}(React.PureComponent);

RadioGroup.displayName = 'AUIRadioGroup';
RadioGroup.defaultProps = {
  onChange: _utils.identity
};
RadioGroup.childContextTypes = {
  auiRadioGroup: _propTypes.object
};
exports.default = RadioGroup;
