// react-aaui/es/form/validation/rules.js
import { PHONE_NUMBER_CONFIG } from '../config';

export var required = function required(val) {
  return !!val;
};
export var regexValue = '^(?=.{0,255}$)[a-z0-9`’!\\^~#{&}\\?=\\$/\\|_%+-]+(\\.[a-z0-9`’!\\^~#{&}\\?=\\$/\\|_%+-]+)*@((([a-z0-9]+)|([a-z0-9]+[a-z0-9-]*[a-z0-9]+))\\.)+[a-z]{2,6}$';
export var email = function email(val) {
  return new RegExp(regexValue, 'i').test(val);
};
export var min = function min() {
  var val = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  var minLen = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  return val.length >= minLen;
};
export var max = function max() {
  var val = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  var maxLen = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  return val.length <= maxLen;
};
export var numeric = function numeric(val) {
  return !isNaN(parseFloat(val)) && isFinite(val);
};
export var url = function url(val) {
  return (/^((((ht|f)tps?):\/\/)|mailto:)[\w-]+(\.[\w-]+)+([\w-.,@?^=%&:/~+#]*[\w-@?^=%&/~+#])?$/i.test(val)
  );
};
export var phone = function phone(val) {
  return PHONE_NUMBER_CONFIG.some(function (phoneNumberPattern) {
    return phoneNumberPattern.test(val);
  });
};
export var regexp = function regexp(val, regex) {
  return val ? new RegExp(regex).test(val) : true;
};

export default {
  required: required,
  email: email,
  min: min,
  max: max,
  numeric: numeric,
  url: url,
  phone: phone,
  regexp: regexp
};
