// react-aaui/lib/form/DateRangeInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _connectForm = require('./connectForm');

var _connectForm2 = _interopRequireDefault(_connectForm);

var _NestedForm = require('./NestedForm');

var _NestedForm2 = _interopRequireDefault(_NestedForm);

var _Form = require('./Form');

var _Form2 = _interopRequireDefault(_Form);

var _DateInput = require('./DateInput');

var _DateInput2 = _interopRequireDefault(_DateInput);

var _types = require('./types');

var _utils = require('../shared/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = (0, _extends3.default)({}, _types.FormFieldAPIPropTypes);
var START_DATE = 'react-aaui.common.dateRangeInput.startDate';
var END_DATE = 'react-aaui.common.dateRangeInput.endDate';

function isBeforeDay(date, day) {
  return date.getTime() < day.getTime();
}

function isAfterDay(date, day) {
  return date.getTime() > day.getTime();
}

var DateRangeInput = function (_PureComponent) {
  (0, _inherits3.default)(DateRangeInput, _PureComponent);

  function DateRangeInput() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, DateRangeInput);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = DateRangeInput.__proto__ || (0, _getPrototypeOf2.default)(DateRangeInput)).call.apply(_ref, [this].concat(args))), _this), _this.getStartDateStatus = function (_ref2) {
      var date = _ref2.date;
      var endDate = _this.props.value.endDate;


      return {
        disabled: endDate && isAfterDay(date, new Date(endDate)) || false
      };
    }, _this.getEndDateStatus = function (_ref3) {
      var date = _ref3.date;
      var startDate = _this.props.value.startDate;


      return {
        disabled: startDate && isBeforeDay(date, new Date(startDate)) || false
      };
    }, _this.setDateClass = function (_ref4) {
      var disabled = _ref4.disabled;
      return disabled ? 'date-picker__disabled' : '';
    }, _this.handleDateClick = function (_ref5) {
      var disabled = _ref5.disabled;
      return !disabled;
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(DateRangeInput, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          value = _props.value,
          l10n = _props.l10n,
          rest = (0, _objectWithoutProperties3.default)(_props, ['value', 'l10n']);
      var startDate = value.startDate,
          endDate = value.endDate;


      return _react2.default.createElement(
        'div',
        { className: 'range-container' },
        _react2.default.createElement(
          _NestedForm2.default,
          (0, _extends3.default)({}, rest, { fields: value }),
          _react2.default.createElement(
            _Form2.default,
            null,
            _react2.default.createElement(_DateInput2.default, {
              name: 'startDate',
              'aria-label': l10n.formatMessage(START_DATE),
              value: startDate,
              getDateStatus: this.getStartDateStatus,
              onClickDate: this.handleDateClick,
              setDateClass: this.setDateClass }),
            _react2.default.createElement(
              'span',
              null,
              'To'
            ),
            _react2.default.createElement(_DateInput2.default, {
              name: 'endDate',
              'aria-label': l10n.formatMessage(END_DATE),
              value: endDate,
              getDateStatus: this.getEndDateStatus,
              onClickDate: this.handleDateClick,
              setDateClass: this.setDateClass })
          )
        )
      );
    }
  }]);
  return DateRangeInput;
}(_react.PureComponent);

DateRangeInput.defaultProps = {
  value: {},
  onChange: _utils.noop
};


DateRangeInput.displayName = 'DateRangeInput';
DateRangeInput.propTypes = propTypes;

exports.default = (0, _connectForm2.default)({})(DateRangeInput);
