// react-aaui/lib/shared/L10nMessage.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _isNil = require('lodash/isNil');

var _isNil2 = _interopRequireDefault(_isNil);

var _injectL10n = require('./injectL10n');

var _injectL10n2 = _interopRequireDefault(_injectL10n);

var _types = require('./types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function L10nMessage(_ref) {
  var id = _ref.id,
      values = _ref.values,
      l10n = _ref.l10n,
      htmlId = _ref.htmlId,
      defaultMessage = _ref.defaultMessage;

  var props = {};
  if (htmlId) {
    props.id = htmlId;
  }
  return _react2.default.createElement(
    'span',
    props,
    l10n.formatMessage(id, values, (0, _isNil2.default)(defaultMessage) ? id : defaultMessage)
  );
}

L10nMessage.propTypes = {
  id: _propTypes.string,
  values: _propTypes.object, // eslint-disable-line
  l10n: _types.aauiL10nShape,
  messageId: _propTypes.string,
  defaultMessage: _propTypes.string
};

exports.default = (0, _injectL10n2.default)()(L10nMessage);
