// react-aaui/lib/alert/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.clear = exports.alert = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _map = require('babel-runtime/core-js/map');

var _map2 = _interopRequireDefault(_map);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _AlertBar = require('./AlertBar');

var _AlertBar2 = _interopRequireDefault(_AlertBar);

var _utils = require('../shared/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var alertMap = new _map2.default();

function clear() {
  alertMap.forEach(function (_ref) {
    var close = _ref.close;
    return close();
  });
  alertMap.clear();
}

function alert(options, element) {
  var props = (0, _extends3.default)({
    onClose: _utils.noop,
    multiple: false,
    inverse: true
  }, options);
  var originalOnClose = props.onClose;
  var target = element || document.createElement('div');

  if (!props.multiple) {
    clear();
  }

  if (!element) {
    target.className = 'global-alert-container';
    document.body && document.body.appendChild(target);
  }

  function close() {
    alertMap.delete(target);
    // Refer to https://facebook.github.io/react/docs/react-dom.html#unmountcomponentatnode
    var unmountResult = _reactDom2.default.unmountComponentAtNode(target);

    !element && unmountResult && target.parentNode && target.parentNode.removeChild(target);

    originalOnClose.apply(undefined, arguments);
  }

  props.onClose = close;
  alertMap.set(target, { close: close });
  _reactDom2.default.render(_react2.default.createElement(_AlertBar2.default, props), target);

  return {
    close: close
  };
}

exports.alert = alert;
exports.clear = clear;
