// react-aaui/es/grid/Col.js
import _toConsumableArray from 'babel-runtime/helpers/toConsumableArray';
import _Object$keys from 'babel-runtime/core-js/object/keys';
import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import * as React from 'react';
import omit from 'lodash/omit';
import classnames from 'classnames';

var defaultProps = {
  span: 12
};

var RESPONSIVE_TIERS = {
  span: 'col',
  sm: 'col-sm',
  md: 'col-md',
  lg: 'col-lg'
};

export default function Col(_ref) {
  var children = _ref.children,
      className = _ref.className,
      responsiveTiers = _ref.responsiveTiers,
      rest = _objectWithoutProperties(_ref, ['children', 'className', 'responsiveTiers']);

  var tiers = _extends({}, RESPONSIVE_TIERS, responsiveTiers);
  var responsiveSettings = _Object$keys(tiers).reduce(function (settings, tier) {
    return rest[tier] ? settings.concat([tiers[tier] + '-' + rest[tier]]) : settings;
  }, []);

  return React.createElement(
    'div',
    _extends({
      className: classnames.apply(undefined, ['col'].concat(_toConsumableArray(responsiveSettings), [className]))
    }, omit(rest, _Object$keys(tiers))),
    children
  );
}

Col.defaultProps = defaultProps;
