// react-aaui/es/grid/Row.js
import _extends from 'babel-runtime/helpers/extends';
import _defineProperty from 'babel-runtime/helpers/defineProperty';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import * as React from 'react';
import classNames from 'classnames';

var defaultProps = {
  fluid: false,
  gutter: true,
  align: 'start',
  justify: 'start',
  style: {}
};

export default function Row(props) {
  var _classNames;

  var fluid = props.fluid,
      justify = props.justify,
      align = props.align,
      className = props.className,
      gutter = props.gutter,
      style = props.style,
      children = props.children,
      rest = _objectWithoutProperties(props, ['fluid', 'justify', 'align', 'className', 'gutter', 'style', 'children']);

  var justifyPrefix = justify === 'between' || justify === 'around' ? 'space-' + justify : justify;

  var boolGutter = typeof gutter === 'boolean';
  var classes = classNames((_classNames = {
    row: !fluid,
    'row-fluid': fluid,
    'row-gutter': boolGutter && !!gutter
  }, _defineProperty(_classNames, 'row-align-' + align, align), _defineProperty(_classNames, 'row-justify-' + justifyPrefix, justify), _classNames), className);

  // prettier-ignore
  var rowStyle = !boolGutter && typeof gutter === 'number' && gutter > 0 ? _extends({
    marginLeft: gutter / -2,
    marginRight: gutter / -2
  }, style) : style;
  var cols = React.Children.map(children, function (col) {
    if (!boolGutter && typeof gutter === 'number' && col && col.props && gutter > 0) {
      return React.cloneElement(col, {
        style: _extends({
          paddingLeft: gutter / 2,
          paddingRight: gutter / 2
        }, col.props.style)
      });
    }
    return col;
  });

  return React.createElement(
    'div',
    _extends({ className: classes, style: rowStyle }, rest),
    cols
  );
}

Row.defaultProps = defaultProps;
