// react-aaui/es/modal/Modal.js
import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import * as React from 'react';
import classnames from 'classnames';

import KEY_CODES from '../shared/keyCodes';
import scopeTab from '../shared/scopeTab';
import { noop } from '../shared/utils';

var Modal = function (_React$PureComponent) {
  _inherits(Modal, _React$PureComponent);

  function Modal() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Modal);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Modal.__proto__ || _Object$getPrototypeOf(Modal)).call.apply(_ref, [this].concat(args))), _this), _this.autoFocus = function () {
      _this.removeTimer = setTimeout(function () {
        _this.focusContent();
        _this.focusAfterRender = false;
      }, 100);
    }, _this.fixBody = function () {
      var bodyStyles = document.body && document.body.style || {};
      var scrollbarWidth = 0;
      if (document.documentElement) {
        scrollbarWidth = window.innerWidth - document.documentElement.clientWidth;
      }

      _this.bodyOverflowX = bodyStyles.overflowX;
      _this.bodyOverflowY = bodyStyles.overflowY;
      _this.bodyPaddingRight = bodyStyles.paddingRight;

      bodyStyles.overflowX = 'hidden';
      bodyStyles.overflowY = 'hidden';

      /* istanbul ignore else */
      if (scrollbarWidth > 0 && document.body) {
        var _getComputedStyle = getComputedStyle(document.body),
            paddingRight = _getComputedStyle.paddingRight;

        bodyStyles.paddingRight = parseInt(paddingRight, 10) + parseInt(scrollbarWidth, 0) + 'px';
        if (_this.root) {
          _this.root.style.paddingRight = bodyStyles.paddingRight;
        }
      }
    }, _this.unfixBody = function () {
      var bodyStyles = document.body && document.body.style || {};
      bodyStyles.overflowX = _this.bodyOverflowX;
      bodyStyles.overflowY = _this.bodyOverflowY;
      bodyStyles.paddingRight = _this.bodyPaddingRight;
      if (_this.root) {
        _this.root.style.paddingRight = '';
      }
    }, _this.focusContent = function () {
      var autoFocusClose = _this.props.autoFocusClose;

      if (!_this.contentHasFocus() && autoFocusClose) {
        _this.closeIcon && _this.closeIcon.focus();
      }
    }, _this.contentHasFocus = function () {
      return document.activeElement === _this.root || _this.root && _this.root.contains(document.activeElement);
    }, _this.handleModalKeydown = function (e) {
      switch (e.keyCode) {
        case KEY_CODES.ESCAPE:
          _this.props.onClose(e);
          break;
        case KEY_CODES.TAB:
          _this.handleTab(e);
          break;
        default:
          break;
      }
    }, _this.handleCloseIconKeydown = function (e) {
      switch (e.keyCode) {
        case KEY_CODES.SPACE:
          _this.props.onClose(e);
          e.preventDefault();
          break;
        default:
          break;
      }
    }, _this.bindRootRef = function (ref) {
      _this.root = ref;
    }, _this.bindBoxRef = function (ref) {
      _this.box = ref;
    }, _this.bindCloseIconRef = function (ref) {
      _this.closeIcon = ref;
    }, _this.render = function () {
      var _this$props = _this.props,
          className = _this$props.className,
          style = _this$props.style,
          title = _this$props.title,
          shown = _this$props.shown,
          onClose = _this$props.onClose,
          children = _this$props.children,
          rest = _objectWithoutProperties(_this$props, ['className', 'style', 'title', 'shown', 'onClose', 'children']);

      var classes = classnames({
        modal: true,
        'is-open': shown
      }, className);

      return React.createElement(
        'div',
        _extends({}, rest, { className: 'modal-wrap', role: 'dialog' }),
        React.createElement(
          'div',
          { className: classes, style: style, ref: _this.bindRootRef, onKeyDown: _this.handleModalKeydown },
          React.createElement(
            'section',
            { className: 'modal-box', ref: _this.bindBoxRef },
            React.createElement(
              'header',
              { className: 'modal-header' },
              React.createElement(
                'h3',
                { className: 'modal-title' },
                title
              ),
              React.createElement('span', { role: 'button', className: 'icon-close modal-close', tabIndex: '0', 'aria-label': 'close-dialog', ref: _this.bindCloseIconRef, onClick: onClose, onKeyDown: _this.handleCloseIconKeydown })
            ),
            children
          )
        ),
        React.createElement('div', { className: 'modal-mask' })
      );
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Modal, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (this.props.shown) {
        this.fixBody();
        this.autoFocus();
      }
      this.focusAfterRender = true;
      this.handleTab = scopeTab(this.root);
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate(_ref2) {
      var shown = _ref2.shown;

      if (shown !== this.props.shown) {
        if (this.props.shown) {
          this.fixBody();
          this.originFocusEl = document.activeElement;
          /* istanbul ignore else */
          if (this.focusAfterRender) {
            this.autoFocus();
          }
        } else {
          this.unfixBody();
          this.originFocusEl && this.originFocusEl.focus();
          this.focusAfterRender = true;
          clearTimeout(this.removeTimer);
        }
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.unfixBody();
    }
  }]);

  return Modal;
}(React.PureComponent);

Modal.defaultProps = {
  shown: false,
  onClose: noop,
  autoFocusClose: true
};
export default Modal;


Modal.displayName = 'AAUIModal';
