// react-aaui/lib/internationalization/LocaleContext.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Consumer = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

exports.Provider = Provider;
exports.consumer = consumer;
exports.withI18n = withI18n;

var _react = require('react');

var React = _interopRequireWildcard(_react);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LocaleContext = React.createContext({});

var getLocale = function getLocale(locales) {
  return function (id) {
    return locales && id && locales[id] || id;
  };
};

var identity = function identity(v) {
  return v;
};

function Provider(_ref) {
  var locales = _ref.locales,
      children = _ref.children,
      rest = (0, _objectWithoutProperties3.default)(_ref, ['locales', 'children']);

  return React.createElement(
    LocaleContext.Provider,
    { value: (0, _extends3.default)({ locales: locales, i18n: getLocale(locales) }, rest) },
    children
  );
}

function consumer(Comp) {
  return function (_ref2) {
    var id = _ref2.id;
    return React.createElement(
      LocaleContext.Consumer,
      null,
      function (_ref3) {
        var _ref3$locales = _ref3.locales,
            locales = _ref3$locales === undefined ? {} : _ref3$locales,
            _ref3$i18n = _ref3.i18n,
            i18n = _ref3$i18n === undefined ? identity : _ref3$i18n;

        return id ? React.createElement(Comp, { value: i18n(id) }) : null;
      }
    );
  };
}

function withI18n(Comp) {
  return function (props) {
    return React.createElement(
      LocaleContext.Consumer,
      null,
      function (_ref4) {
        var _ref4$i18n = _ref4.i18n,
            i18n = _ref4$i18n === undefined ? identity : _ref4$i18n;
        return React.createElement(Comp, (0, _extends3.default)({ i18n: i18n }, props));
      }
    );
  };
}

var Consumer = LocaleContext.Consumer;
exports.Consumer = Consumer;
