// react-aaui/lib/Checkbox/CheckboxGroup.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _propTypes = require('prop-types');

var _Checkbox = require('./Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _utils = require('../shared/utils');

require('./CheckboxGroup.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CheckboxGroup = function (_React$PureComponent) {
  (0, _inherits3.default)(CheckboxGroup, _React$PureComponent);

  function CheckboxGroup(props) {
    (0, _classCallCheck3.default)(this, CheckboxGroup);

    var _this = (0, _possibleConstructorReturn3.default)(this, (CheckboxGroup.__proto__ || (0, _getPrototypeOf2.default)(CheckboxGroup)).call(this, props));

    _this.onChange = function (e, _ref) {
      var optionValue = _ref.value;

      // Should construct one new array for re-rendering
      var value = [].concat((0, _toConsumableArray3.default)(_this.state.value));
      var onChange = _this.props.onChange;

      var index = value.indexOf(optionValue);

      if (index === -1) {
        // If NOT found, then add it
        value.push(optionValue);
      } else {
        // If found, then delete it
        value.splice(index, 1);
      }

      if (!('value' in _this.props)) {
        _this.setState({
          value: value
        });
      }

      onChange(value);
    };

    _this.state = {
      value: props.value || props.defaultValue || []
    };
    return _this;
  }

  (0, _createClass3.default)(CheckboxGroup, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        auiCheckboxGroup: {
          onChange: this.onChange,
          value: this.state.value,
          disabled: this.props.disabled,
          size: this.props.size,
          name: this.props.name
        }
      };
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if ('value' in nextProps) {
        this.setState({
          value: nextProps.value
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var data = this.props.data;
      var children = this.props.children;


      if (data) {
        children = data.map(function (_ref2, index) {
          var text = _ref2.text,
              value = _ref2.value,
              rest = (0, _objectWithoutProperties3.default)(_ref2, ['text', 'value']);
          return React.createElement(
            _Checkbox2.default,
            (0, _extends3.default)({ key: index, value: value }, rest),
            text
          );
        });
      }

      return React.createElement(
        'fieldset',
        null,
        React.createElement('legend', null),
        children
      );
    }
  }]);
  return CheckboxGroup;
}(React.PureComponent);

CheckboxGroup.displayName = 'AUICheckboxGroup';
CheckboxGroup.defaultProps = {
  onChange: _utils.identity
};
CheckboxGroup.childContextTypes = {
  auiCheckboxGroup: _propTypes.object
};
exports.default = CheckboxGroup;
