// react-aaui/lib/Checkbox/CheckboxComponent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _utils = require('../shared/utils');

var _L10nMessage = require('../shared/L10nMessage');

var _L10nMessage2 = _interopRequireDefault(_L10nMessage);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CHECKBOX = 'react-aaui.common.checkbox';

var Checkbox = function (_React$PureComponent) {
  (0, _inherits3.default)(Checkbox, _React$PureComponent);

  function Checkbox(props) {
    (0, _classCallCheck3.default)(this, Checkbox);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Checkbox.__proto__ || (0, _getPrototypeOf2.default)(Checkbox)).call(this, props));

    _initialiseProps.call(_this);

    var checked = 'checked' in props ? props.checked : props.defaultChecked;

    _this.state = {
      checked: !!checked
    };
    return _this;
  }

  (0, _createClass3.default)(Checkbox, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      (0, _defineProperties2.default)(this, {
        checked: {
          get: function get() {
            return this.input.checked;
          },
          set: function set(v) {
            this.input.checked = !!v;
            this.setState({
              checked: !!v
            });
          }
        }
      });
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if ('checked' in nextProps) {
        this.setState({
          checked: !!nextProps.checked
        });
      }
    }

    // For WCAG 2.0, a form label cannot contain no context.
    // Related ticket: https://jirafnd.dev.activenetwork.com/browse/AWF-186

  }, {
    key: 'renderDefaultText',
    value: function renderDefaultText(ariaLabelId) {
      return React.createElement(
        'span',
        { id: ariaLabelId, style: { display: 'none' } },
        React.createElement(_L10nMessage2.default, { id: CHECKBOX })
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          toggle = _props.toggle,
          size = _props.size,
          children = _props.children,
          className = _props.className,
          style = _props.style,
          disabled = _props.disabled,
          ariaLabelId = _props.ariaLabelId,
          rest = (0, _objectWithoutProperties3.default)(_props, ['toggle', 'size', 'children', 'className', 'style', 'disabled', 'ariaLabelId']);
      var checked = this.state.checked;

      var checkboxWrapperClasses = (0, _classnames2.default)({
        'checkbox-wrapper': true,
        toggle: toggle,
        'toggle--empty': !children && toggle
      }, className);
      var checkboxClasses = (0, _classnames2.default)((0, _defineProperty3.default)({
        checkbox: !toggle,
        'checkbox--checked': !toggle && checked,
        'checkbox--disabled': !toggle && disabled
      }, 'checkbox--' + (size || ''), size || ''));

      return React.createElement(
        'label',
        { className: checkboxWrapperClasses, style: style },
        React.createElement(
          'span',
          { className: checkboxClasses },
          React.createElement('input', (0, _extends3.default)({
            'aria-labelledby': ariaLabelId,
            ref: this.setWrappedComponentInstance,
            className: 'checkbox__input',
            disabled: disabled,
            type: 'checkbox',
            'aria-hidden': true
          }, rest, {
            checked: checked,
            onChange: this.handleChange })),
          toggle ? React.createElement('span', { className: 'toggle__text' }) : React.createElement('span', {
            className: 'checkbox__inner',
            role: 'checkbox',
            'aria-disabled': disabled,
            'aria-checked': checked }),
          children ? React.createElement(
            'span',
            { id: ariaLabelId, className: 'checkbox__text' },
            children
          ) : this.renderDefaultText(ariaLabelId)
        )
      );
    }
  }]);
  return Checkbox;
}(React.PureComponent);

Checkbox.displayName = 'AUICheckbox';
Checkbox.defaultProps = {
  onChange: _utils.identity,
  defaultChecked: false,
  ariaLabelId: undefined
};

var _initialiseProps = function _initialiseProps() {
  var _this2 = this;

  this.setWrappedComponentInstance = function (input) {
    _this2.input = input;
  };

  this.triggerOnChange = function (e) {
    var onChange = _this2.props.onChange;
    var checked = e.target.checked;


    if (!('checked' in _this2.props)) {
      _this2.setState({
        checked: checked
      });
    }

    // If you want to access the event properties in an asynchronous way,
    // you should call `event.persist()`` on the event,
    // which will remove the synthetic event from the pool
    // and allow references to the event to be retained by user code.
    e.persist && e.persist();

    // The `value` prop is consumed by `CheckboxGroup`
    onChange(e, { value: _this2.props.value });
  };

  this.handleChange = function (e) {
    var disabled = _this2.props.disabled;


    if (disabled) {
      return;
    }

    _this2.triggerOnChange(e);
  };
};

exports.default = Checkbox;
