// react-aaui/lib/loading/LoadingCounter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LoadingCounter = function () {
  function LoadingCounter() {
    (0, _classCallCheck3.default)(this, LoadingCounter);
    this.count = 0;
  }

  (0, _createClass3.default)(LoadingCounter, [{
    key: "add",
    value: function add() {
      this.count += 1;
    }
  }, {
    key: "dec",
    value: function dec() {
      this.count -= 1;
    }
  }, {
    key: "isEmpty",
    value: function isEmpty() {
      return this.count <= 0;
    }
  }, {
    key: "clear",
    value: function clear() {
      this.count = 0;
    }
  }]);
  return LoadingCounter;
}();

exports.default = LoadingCounter;
