// react-aaui/es/alert/util.js
import _extends from 'babel-runtime/helpers/extends';
import _Map from 'babel-runtime/core-js/map';
import React from 'react';
import ReactDOM from 'react-dom';

import AlertBar from './AlertBar';
import { noop } from '../shared/utils';

var alertMap = new _Map();

function clear() {
  alertMap.forEach(function (_ref) {
    var close = _ref.close;
    return close();
  });
  alertMap.clear();
}

function alert(options, element) {
  var props = _extends({
    onClose: noop,
    multiple: false,
    inverse: true
  }, options);
  var originalOnClose = props.onClose;
  var target = element || document.createElement('div');

  if (!props.multiple) {
    clear();
  }

  if (!element) {
    target.className = 'global-alert-container';
    document.body && document.body.appendChild(target);
  }

  function close() {
    alertMap.delete(target);
    // Refer to https://facebook.github.io/react/docs/react-dom.html#unmountcomponentatnode
    var unmountResult = ReactDOM.unmountComponentAtNode(target);

    !element && unmountResult && target.parentNode && target.parentNode.removeChild(target);

    originalOnClose.apply(undefined, arguments);
  }

  props.onClose = close;
  alertMap.set(target, { close: close });
  ReactDOM.render(React.createElement(AlertBar, props), target);

  return {
    close: close
  };
}

export { alert, clear };
