// react-aaui/lib/tabs/TabsContainer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _types = require('../shared/types');

var _keyCodes = require('../shared/keyCodes');

var _keyCodes2 = _interopRequireDefault(_keyCodes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var handleKeyDown = function handleKeyDown(e) {
  e.persist();
  switch (e.keyCode) {
    case _keyCodes2.default.LEFTARROW:
    case _keyCodes2.default.RIGHTARROW:
      e.stopPropagation();
      break;
    default:
      break;
  }
};

var TabsContainer = function TabsContainer(_ref) {
  var className = _ref.className,
      style = _ref.style,
      name = _ref.name,
      children = _ref.children,
      getSelected = _ref.auiTabsAPI.getSelected;

  var classes = (0, _classnames2.default)({
    'u-hidden': name !== getSelected()
  }, className);

  return React.createElement(
    'div',
    {
      className: classes,
      style: style,
      role: 'tabpanel',
      'aria-hidden': name !== getSelected(),
      onKeyDown: handleKeyDown },
    children
  );
};

TabsContainer.displayName = 'AUITabsContainer';

exports.default = TabsContainer;
