// react-aaui/es/radio/RadioComponent.js
import _extends from 'babel-runtime/helpers/extends';
import _defineProperty from 'babel-runtime/helpers/defineProperty';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _Object$defineProperties from 'babel-runtime/core-js/object/define-properties';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import * as React from 'react';
import classNames from 'classnames';

import { identity } from '../shared/utils';

var Radio = function (_React$PureComponent) {
  _inherits(Radio, _React$PureComponent);

  function Radio(props) {
    _classCallCheck(this, Radio);

    var _this = _possibleConstructorReturn(this, (Radio.__proto__ || _Object$getPrototypeOf(Radio)).call(this, props));

    _initialiseProps.call(_this);

    var checked = 'checked' in props ? props.checked : props.defaultChecked;

    _this.state = {
      checked: !!checked
    };
    return _this;
  }

  _createClass(Radio, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      _Object$defineProperties(this, {
        checked: {
          get: function get() {
            return this.input.checked;
          },
          set: function set(v) {
            this.input.checked = !!v;
            this.setState({
              checked: !!v
            });
          }
        }
      });
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if ('checked' in nextProps) {
        this.setState({
          checked: !!nextProps.checked
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      /* eslint-disable no-unused-vars */
      var _props = this.props,
          size = _props.size,
          children = _props.children,
          className = _props.className,
          style = _props.style,
          disabled = _props.disabled,
          defaultChecked = _props.defaultChecked,
          ariaLabelId = _props.ariaLabelId,
          rest = _objectWithoutProperties(_props, ['size', 'children', 'className', 'style', 'disabled', 'defaultChecked', 'ariaLabelId']);

      var checked = this.state.checked;

      var radioWrapperClasses = classNames({
        'radio-wrapper': true
      }, className);
      var radioClasses = classNames(_defineProperty({
        radio: true,
        'radio--checked': checked,
        'radio--disabled': disabled
      }, 'radio--' + size, size));

      /* eslint-disable jsx-a11y/label-has-for */
      return React.createElement(
        'label',
        { className: radioWrapperClasses, style: style },
        React.createElement(
          'span',
          { className: radioClasses },
          React.createElement('input', _extends({
            ref: this.setWrappedComponentInstance,
            className: 'radio__input',
            disabled: disabled,
            type: 'radio',
            'aria-hidden': true,
            'aria-labelledby': ariaLabelId
          }, rest, {
            checked: checked,
            onChange: this.handleChange })),
          React.createElement('span', {
            className: 'radio__inner',
            role: 'radio',
            'aria-disabled': disabled,
            'aria-checked': checked }),
          React.createElement(
            'span',
            { id: ariaLabelId },
            children
          )
        )
      );
    }
  }]);

  return Radio;
}(React.PureComponent);

Radio.defaultProps = {
  size: '',
  defaultChecked: false,
  onChange: identity,
  ariaLabelId: undefined
};

var _initialiseProps = function _initialiseProps() {
  var _this2 = this;

  this.setWrappedComponentInstance = function (input) {
    _this2.input = input;
  };

  this.triggerOnChange = function (e) {
    var onChange = _this2.props.onChange;
    var checked = e.target.checked;


    if (!('checked' in _this2.props)) {
      _this2.setState({
        checked: checked
      });
    }

    // If you want to access the event properties in an asynchronous way,
    // you should call `event.persist()`` on the event,
    // which will remove the synthetic event from the pool
    // and allow references to the event to be retained by user code.
    e.persist && e.persist();

    onChange(e);
  };

  this.handleChange = function (e) {
    var disabled = _this2.props.disabled;


    if (disabled) {
      return;
    }

    _this2.triggerOnChange(e);
  };
};

export default Radio;
