// react-aaui/es/radio/Radio.js
import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _Object$defineProperties from 'babel-runtime/core-js/object/define-properties';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import * as React from 'react';
import { object } from 'prop-types';

import RadioComponent from './RadioComponent';
import { identity } from '../shared/utils';

var Radio = function (_React$Component) {
  _inherits(Radio, _React$Component);

  function Radio() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Radio);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Radio.__proto__ || _Object$getPrototypeOf(Radio)).call.apply(_ref, [this].concat(args))), _this), _this.setRef = function (radioComponent) {
      _this.radioComponent = radioComponent;
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Radio, [{
    key: 'componentDidMount',


    // Support legacy API for `this.refs.checkbox.checked`
    // It's not recommended because it breaks component encapsulation and
    // it violates the unidirectional data flow for React
    value: function componentDidMount() {
      _Object$defineProperties(this, {
        checked: {
          get: function get() {
            return this.radioComponent.checked;
          },
          set: function set(v) {
            this.radioComponent.checked = !!v;
          }
        }
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          rest = _objectWithoutProperties(_props, ['children']);

      var auiRadioGroup = this.context.auiRadioGroup;

      // Proxy `onChange`, `checked` and `disabled` to `auiRadioGroup`

      if (auiRadioGroup) {
        rest.onChange = auiRadioGroup.onChange;
        rest.checked = this.props.value === auiRadioGroup.value;
        rest.disabled = this.props.disabled || auiRadioGroup.disabled;
        rest.size = this.props.size || auiRadioGroup.size;
        rest.name = this.props.name || auiRadioGroup.name;
      }

      return React.createElement(
        RadioComponent,
        _extends({ ref: this.setRef }, rest),
        children
      );
    }
  }]);

  return Radio;
}(React.Component);

Radio.displayName = 'AUIRadio';
Radio.defaultProps = {
  onChange: identity
};
Radio.contextTypes = {
  auiRadioGroup: object
};
export default Radio;
