// react-aaui/es/ProgressSteps.js
import _defineProperty from 'babel-runtime/helpers/defineProperty';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent } from 'react';
import { string, object, array } from 'prop-types';
import classNames from 'classnames';

import * as da from './shared/data-access';

var ProgressSteps = function (_PureComponent) {
  _inherits(ProgressSteps, _PureComponent);

  function ProgressSteps() {
    _classCallCheck(this, ProgressSteps);

    return _possibleConstructorReturn(this, (ProgressSteps.__proto__ || _Object$getPrototypeOf(ProgressSteps)).apply(this, arguments));
  }

  _createClass(ProgressSteps, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          className = _props.className,
          style = _props.style,
          size = _props.size,
          steps = _props.steps;

      var classes = classNames(_defineProperty({
        'progress-steps': true
      }, 'progress-steps--' + size, size), className);

      return React.createElement(
        'ul',
        { className: classes, style: style },
        steps.map(function (step) {
          return React.createElement(
            'li',
            {
              key: da.get(step, 'text'),
              className: 'progress-steps__step ' + da.get(step, 'status') },
            React.createElement(
              'span',
              { className: 'progress-steps__step-text' },
              da.get(step, 'text')
            )
          );
        })
      );
    }
  }]);

  return ProgressSteps;
}(PureComponent);

ProgressSteps.displayName = 'AAUIProgressSteps';
ProgressSteps.propTypes = {
  className: string,
  size: string,
  steps: array, // eslint-disable-line
  style: object // eslint-disable-line
};
export default ProgressSteps;
